swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1292,7 +1403,111 @@ string ppReprType(DataFlowType t) { none() }
  * a node of type `t1` to a node of type `t2`.
  */
 pragma[inline]
-predicate compatibleTypes(DataFlowType t1, DataFlowType t2) { any() }
+predicate compatibleTypes(DataFlowType t1, DataFlowType t2) {
+  subType(t1) = subType(t2)
+  or
+  exists(BoundGenericType bound1, BoundGenericType bound2 |
+    stripType(t1.asType()).(BoundGenericType) = bound1 and
+    stripType(t2.asType()).(BoundGenericType) = bound2 and
+    bound1.getDeclaration() = bound2.getDeclaration() and
+    forall(int index | index in [0 .. bound1.getNumberOfArgTypes() - 1] |
+      isAnyOrUnboundType(bound1.getArgType(index).getCanonicalType()) or
+      isAnyOrUnboundType(bound2.getArgType(index).getCanonicalType())
+    )
+  )
+  or
+  exists(TupleType tuple1, TupleType tuple2 |
+    stripType(t1.asType()) = tuple1 and
+    stripType(t2.asType()) = tuple2 and
+    compatibleTuples(tuple1, tuple2)
```

Comment:
Similar question here: I _think_ that `compatibleTuples` holds when there is a common index in which both tuple types is either the any type or an unbound type. When no such index exists, does that mean they're compatible because of the first condition in `compatibleTypes`?